networkit/cpp/centrality/ComplexPaths.cpp

```diff
@@ -0,0 +1,205 @@
+    if (mode == Mode::singleNode && start == none)
+        throw std::runtime_error("complexPathAlgorithm: provide node != none in Mode::singleNode.");
+    if (mode == Mode::allNodes && start != none)
+        throw std::runtime_error("complexPathAlgorithm: start node ignored in Mode::allNodes.");
+}
+
+void ComplexPathAlgorithm::normalize() {
+    if (mode != Mode::allNodes) {
+        WARN("complexPathAlgorithm: no normalization in Mode::singleNode.");
+        return;
+    }
+    normPaths = true;
+}
+
+void ComplexPathAlgorithm::run() {
+    if (mode == Mode::singleNode) {
+        complexPathGraph = complexPathsGraph(start, threshold, &adopters);
+    } else {
+        complexPathsLengths = complexPathLength(threshold);
+    }
+    hasRun = true;
+}
+
+std::vector<double> ComplexPathAlgorithm::getPLci() {
+    if (mode != Mode::allNodes)
+        WARN("complexPathAlgorithm[getPLci]: no results in Mode::singleNode.");
+    assureFinished();
+    return complexPathsLengths;
+}
+
+Graph ComplexPathAlgorithm::getComplexGraph() {
+    if (mode != Mode::singleNode)
+        WARN("complexPathAlgorithm[getComplexGraph]: no results in Mode::allNodes.");
+    assureFinished();
+    return complexPathGraph;
+}
+
+std::vector<node> ComplexPathAlgorithm::getAdopters() {
+    if (mode != Mode::singleNode)
+        WARN("complexPathAlgorithm[getAdopters]: no results in Mode::allNodes.");
+    assureFinished();
+    return adopters;
+}
+
+static void addNewEdge(Graph &g, node u, node v) {
+    if (g.hasEdge(u, v))
+        return;
+    g.addEdge(u, v);
+}
+
+std::vector<node> ComplexPathAlgorithm::generateSeeds(node seed, const Graph &g, count threshold) {
+    // activate threshold-1 random neighbors of seed:
+    auto seedsNeeded = threshold - 1;
+
+    std::vector<node> seeds;
+    seeds.reserve(4);
+
+    auto neighbors = Graph::NeighborRange<>(g, seed);
+
+    auto randomGen = Aux::Random::getURNG();
+    std::sample(neighbors.begin(), neighbors.end(), std::back_inserter(seeds), seedsNeeded,
+                randomGen);
+    auto needMore = seedsNeeded - seeds.size();
+    if (needMore > 0) {
+        WARN("complexPathsGraph: ", seed, ": not enough direct neighbors.");
+
+        // collect all nodes at distance 2 before sampling to establish equal distribution
+        std::vector<node> indirNeighbors;
+
+        for (node u : neighbors) {
+            auto indirectNeighbors = Graph::NeighborRange<>(g, u);
+            for (node in : indirectNeighbors) {
+                if (in != seed) {
+                    indirNeighbors.push_back(in);
+                }
+            }
+        }
+        std::sample(indirNeighbors.begin(), indirNeighbors.end(), std::back_inserter(seeds),
+                    needMore, randomGen);
+    }
+    return seeds;
+}
+
+Graph ComplexPathAlgorithm::complexPathsGraph(node seed, count threshold,
+                                              std::vector<node> *adopters) {
```

Comment:
Why don't we use a reference?
```suggestion
                                              std::vector<node> &adopters) {
```

Reply:
As far as i can tell, because of line 191, where we call complexPathGraphs using the nullpointer. (is there a way around this?)

Reply:
I see. Then let's keep the pointer.